### ISSUE TITLE
Delaying input for multi-key combinations

### ISSUE THREAD

**Issue by DimmyJing** [VSCodeVim/Vim#3082]
# Is your feature request related to a problem? Please describe.

I'm always frustrated when I type part of a key combination and the line moves along with the first key. For example, I have remapped jk to <esc> and the editor would always type out the j, deleting it after I type the k.

# Describe the solution you'd like
The solution by NeoVim is to replace the next character with the character you're typing, and inserting the character after it times out. Personally, I find it to be a much cleaner solution.

# Describe alternatives you've considered
I have no idea for the possibility for this feature, but just delaying the actual insertion for the key would be enough. 

# Additional context
What VSCode does
![whathappens](https://user-images.githubusercontent.com/16280574/46259349-ef63ba00-c49d-11e8-9141-625244465612.gif)
What VSCode should do
![whatshouldhappen](https://user-images.githubusercontent.com/16280574/46259378-508b8d80-c49e-11e8-88ae-b4448b36bf1c.gif)

**Comment by mauricecruz** [comment]
Chiming in here to see if any discovery was made towards this issue. I also remapped my keys similarly. The character insertion ends up in the history buffer which makes undo'ing actions noisy. 😞

**Comment by DimmyJing** [comment]
I've looked through the code and I think I found a possible solution to this problem.

Basically the character after the cursor would be overridden by the current key if the `isPotentialRemap` property of class `Remappers` is true, and the cursor would be moved back one step. So mappings such as jk to <esc> would be `|test -> |jest -> |test`

Every consecutive character would also override the character right after the cursor and move back the cursor by one step. So, for example, three-letter mappings such as abc to <esc> would be `|test -> |aest -> |best -> |test`

Failed remappings should be like `|test -> |jest -> jk|test` and `|test -> |aest -> |best -> abc|test`

To implement this, I think `vimState` can add a new char field called something like `overwrittenChar` that holds the character that was overwritten.

Then the first part of the `handleRemapping` function in `remapper.ts` could be modified to only replace the letter after the cursor position to `overwrittenChar`

And there could be another function somewhere called `remapFailed` that is called whenever the next character sets `isPotentialRemap` to false or the current key has timeout: the current time - `vimState.lastKeyPressedTimestamp` is greater than `configuration.timeout`

What `remapFailed` does is basically replace the character after the cursor to `overwrittenChar`, add the letters in `vimState.recordedState.commandList` before the cursor.

I would have implemented it myself, but the only thing is that `remapFailed` needs `setTimeout` to function correctly (to cancel the remap if the time passed is greater than timeout), but I'm not sure how to implement that to modify vimState. It would be great if someone could implement this or point me to the right direction to solve the problem.
